Timeout setting for Opensearch and Elasticsearch

z-mrozu opened the pull request:
**Description:**

Added timeout setting in Opensearch & Elasticsearch config which should only be relevant if user sets "ES_TIMEOUT"

**PR Checklist:**

- [x] Code is formatted and linted (run `pre-commit run --all-files`)
- [x] Tests pass (run `make test`)
- [x] Documentation has been updated to reflect changes, if applicable
- [x] Changes are added to the changelog

jonhealy1 commented on stac_fastapi/elasticsearch/stac_fastapi/elasticsearch/config.py:
```diff
@@ -56,6 +56,10 @@ def _es_config() -> Dict[str, Any]:
 
+    # Include timeout setting if set
+    if timeout := os.getenv("ES_TIMEOUT"):
+        config["timeout"] = timeout
```
Hi. Nice addition. Should this be `request_timeout`? https://www.elastic.co/guide/en/elasticsearch/client/python-api/8.18/config.html#timeouts

jonhealy1 commented:
@z-mrozu Can you add an entry here for the Readme? https://github.com/stac-utils/stac-fastapi-elasticsearch-opensearch?tab=readme-ov-file#configuration-reference

z-mrozu commented on stac_fastapi/opensearch/stac_fastapi/opensearch/config.py:
```diff
@@ -53,6 +53,10 @@ def _es_config() -> Dict[str, Any]:
 
+    # Include timeout setting if set
+    if timeout := os.getenv("ES_TIMEOUT"):
+        config["timeout"] = timeout
```
It should be, from [OpenSearch Client](https://opensearch-project.github.io/opensearch-py/api-ref/clients/opensearch_client.html#opensearchpy.OpenSearch) parameters:
"kwargs ([Any](https://docs.python.org/3/library/typing.html#typing.Any)) – any additional arguments will be passed on to the [Transport](https://opensearch-project.github.io/opensearch-py/api-ref/transport.html#opensearchpy.Transport) class and, subsequently, to the [Connection](https://opensearch-project.github.io/opensearch-py/api-ref/connection.html#opensearchpy.Connection) instances." 
and then [Connection](https://opensearch-project.github.io/opensearch-py/api-ref/connection.html#opensearchpy.Connection) has the timeout parameter:
"timeout ([int](https://docs.python.org/3/library/functions.html#int)) – default timeout in seconds (float, default: 10)"

jonhealy1 replied:
Thanks for looking into this, I was going through the opensearch-py code a little and this makes sense for sure.

jonhealy1 commented:
@z-mrozu I think maybe we say something general, because we don't actually set a default. We could leave it blank or say something like - uses db client default?

z-mrozu commented:
@jonhealy1 I added es_timeout to the config reference and went for "DB client default" in Default column. I also cleaned up the formatting of the table a bit, hope that's okay

jonhealy1 commented:
This is from ai so I know it's questionable information but it is something that I am wondering about - are `timeout` and `request_timeout` two different things?

```
Why timeout and request_timeout Differ:

Elasticsearch (elasticsearch-py): timeout sets the connection timeout (e.g., 10s for establishing a connection), while request_timeout sets the total request timeout (e.g., 30s for connection, sending, processing, response) for all API calls.

OpenSearch (opensearch-py): Only timeout is available at client initialization, controlling connection timeout. A tuple timeout=(connect_timeout, read_timeout) can approximate request_timeout by setting read_timeout for the response phase, but it’s HTTP-layer (less precise) and applies globally.

Difference: timeout is for connection establishment; request_timeout (Elasticsearch) or read_timeout (OpenSearch) covers the entire request lifecycle.

Simplest Solution
Use a single SEARCH_TIMEOUT environment variable (e.g., 30) to set request_timeout for Elasticsearch or read_timeout (with fixed connect_timeout=10) for OpenSearch in your FastAPI app.
```

I can try to verify this info when I have time. Are you a Elasticsearch or Opensearch User?

z-mrozu commented:
Mostly Opensearch user, I think `request_timeout` used to be named `timeout` in ES client at some point (based on the deprecation warning) but I'm unsure if OS `timeout` and ES `timeout` were implemented in the same way or not. At least in Opensearch the setting seems to control connection timeout - as in if you can connect with OS within set `timeout` but for example you are making a sizeable search request that takes a while to actually download, maybe even longer than the set `timeout`, you're still in the clear and the search request doesn't throw a timeout error

z-mrozu commented:
In earlier versions of [ES (7.x)](https://elasticsearch-py.readthedocs.io/en/7.x/api.html#elasticsearch.Elasticsearch) it _seems_ that the `timeout` param was the same - as in it was a kwarg passed to [Connection](https://elasticsearch-py.readthedocs.io/en/7.x/transports.html?highlight=connection#elasticsearch.connection.Connection) with default value set as 10, but again I'm unsure about how and why it was changed to `request_timeout`

jonhealy1 commented:
The logic here - I guess - is that the Requests library will accept a tuple and there is nothing in the opensearch code that will prevent this? https://requests.readthedocs.io/en/latest/user/advanced/#timeouts

Not my words: 

The timeout=(connect_timeout, read_timeout) tuple works in opensearch-py because the RequestsHttpConnection class (in opensearchpy/connection/http_requests.py) passes the timeout parameter directly to the requests library’s Session.send method, which supports a tuple for setting separate connection and read timeouts. The requests library interprets the tuple as (connect_timeout, read_timeout), applying the first value to establish the connection and the second to wait for the server’s response, ensuring both timeouts are enforced for all API calls despite the misleading Optional[Union[int, float]] type hint.

jonhealy1 reviewed (approved):
Thanks @z-mrozu!
